File: 시즌 11/중/kimryungyo/other_languages/day_8.java
import java.util.Stack;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        String nums_string = br.readLine();
        String[] nums_str = nums_string.split(" ");
        int[] stones = new int[N];;
        for (int i = 0; i < nums_str.length; i++) {
            stones[i] = Integer.parseInt(nums_str[i]);
        }

        int S = Integer.parseInt(br.readLine());

        boolean[] visited = new boolean[N];
        visited[S - 1] = true;

        int next, jump, count = 0;
        Stack<Integer> stack = new Stack<>();
        stack.push(S - 1);

        while (!stack.isEmpty()) {
            int pos = stack.pop();
            count++;

            jump = stones[pos];

            next = pos - jump;
            if (next >= 0 && next < N && !visited[next]) {
                visited[next] = true;
                stack.push(next);
            }

            next = pos + jump;
            if (next >= 0 && next < N && !visited[next]) {
                visited[next] = true;
                stack.push(next);
            }

        }

        System.out.println(count);
    }
}

File: 시즌 11/중/kimryungyo/other_languages/day_8.c
// DFS보단 BFS를 선호하는데 큐는 구현이 복잡할 거 같아서 DFS로..

#include <stdio.h>

int main(void) {
    int N, S, stones[100001];
    scanf("%d", &N);
    for (int i = 0; i < N; i++) { scanf("%d", &stones[i]); }
    scanf("%d", &S);

    int stack[1024], visited[100001] = {0};
    int count, pointer, pos, next, jump;
    
    count = 0;
    
    stack[0] = S - 1;
    pointer = 1;
    
    while (pointer) {
        pos = stack[pointer - 1];
        pointer--;
        count++;

        jump = stones[pos];

        next = pos - jump;
        if ((0 <= next) && (next < N) && !visited[next]) {
            visited[next] = 1;
            stack[pointer] = next;
            pointer++;
        }

        next = pos + jump;
        if ((0 <= next) && (next < N) && !visited[next]) {
            visited[next] = 1;
            stack[pointer] = next;
            pointer++;
        }
    }

    printf("%d", count);
    
}